### ISSUE TITLE
Ability of changeing state twice in on action | dispatching action inside another | disposing actions in async way

### ISSUE THREAD

**Issue by ConductedClever** [aurelia/store#85]
**I'm submitting a feature request**

* **Library Version:**
1.1.0

* **Operating System:**
Windows 10

* **Node Version:**
11.2.0

* **NPM Version:**
6.4.1

* **Aurelia CLI OR JSPM OR Webpack AND Version**
CLI 1.0.0-beta.6

* **Browser:**
all

* **Language:**
ESNext

**Current behavior:**
In our company an skeleton (template) is developed to be used in many projects. The main idea of this skeleton is to use aurelia-store as the main manager for data, services and etc.

The solution is this way: In viewmodel instead of doing main jobs or extracting the main job to a service, we dispatch one or more actions and inside it all jobs get done. For example the action gets some params from viewmodel, sends them to a remote server (web api) and then change the state.

Using this skeleton we have 3 following issues:

1. The very main issue is that dispatching actions are scheduled in an async queue. So that, when there are many components side by side, and there are many action dispatches inside them, all of them are queued to be done one after another and not async to each other.
2. The second issue is that, as the state can be changed one time along each action dispatch, we can not do some related jobs aside each other. For example assume that I want to display a loading indicator before calling web api and hide it after data retrieved. For making the mask disappear, I can change the loading state at the time data is received and I am putting the data into store, but for showing loading, I need to change the state once before data retrieved and once after that (which is impossible).
3. And the last one is, dispatching one action inside another one. Additional to second issue, maybe I want to wrap the first state update into another action and dispatch the second one inside first one. But in the actions, there is no access to store, and hence there is no access to dispatch.

**Expected/desired behavior:**
1. Move the async queue behaviour into the state changing part. So that, the other lines of code will run asynchronously and the state update part being queued in an async queue.
2. Like the observables in rxjs, let the developer to send next state more that one time. For example instead of returning next state, call a predefined function named `nextState(nextState);`.
3. Let an action see the store as its `this` scope and call dispatch function on it.

* **What is the motivation / use case for changing the behavior?**
The motivation of these features is to let action replace the use cases of service (login layer) and be more generic in using purposes.

* **Additional comments**
There is another solution to not facing these issues. The solution will be to place a service layer before action dispatch. So that inside the service as data update in store is needed, a dispatch will be called.
But in our skeleton, we want to reduce the layers count and remove service layer as it is not necessary.

**Comment by timfish** [comment]
> 1. The very main issue is that dispatching actions are scheduled in an async queue. ...queued to be done one after another and not async to each other.

Executing the actions in order, without them overlapping is fundamental to how state containers work. Each action must be passed the previous state and that's not available until the previous action has completed.

If you have multiple actions running at the same time, what state would you pass each when you start them and what would you do with their most likely differing resulting states?

> 2. Like the observables in rxjs, let the developer to send next state more that one time

Again, for the same reasons above, an action can't modify the state more than once because the state might have been modified since the first time it resulted. 

> 3. ...dispatching one action inside another one...  ...Let an action see the store as its this scope and call dispatch function on it....

You can dispatch an action without a direct reference to the store using [`dispatchify`](https://aurelia.io/docs/plugins/store#using-the-dispatchify-higher-order-function).

**Comment by zewa666** [comment]
@timfish did properly summarise why it's a bad idea to do what you proposed ... in general. Now there for whatever reason might be cases where you really want to do that and ignore potential sideeffects. Again this is definitely something that might go wrong but you willingly ignore.

So by looking at the store implementation you can see that dispatch does immediately call the [internal Dispatchqueue](https://github.com/aurelia/store/blob/master/src/store.ts#L117-L122). The handleQueue method than later calls [internalDispatch](https://github.com/aurelia/store/blob/master/src/store.ts#L141) and this is exactly the thing where you can jump in to bypass the queue.

So a simple example could be

```typescript
const myFirstAction = (state) => {
  return Object.assign({}, state, { your: "Modification" }
}

const mySecondAction = (state) => {
  return Object.assign({}, state, { your: "Modification" }
}

class App {
  constructor(store: Store<State>) {
    store.registerAction("A", myFirstAction);
    store.registerAction("B", mySecondAction);

   // workaround the queue
   (store as any).internalDispatch(myFirstAction);
   (store as any).internalDispatch(mySecondAction);
  }
}
```

You can also go crazy and dispatch actions from actions if that is really what you want to do

```
const mySecondAction = (state) => {
  return Object.assign({}, state, { your: "Modification" }
}

class App {
  constructor(store: Store<State>) {
    const mySecondAction = async (state) => {
     const modification = await myService.asyncFunction();
     const firstModification = Object.assign({}, state, { your: modification }
     (store as any).internalDispatch(mySecondAction);

    return firstModification;
   }
    store.registerAction("A", myFirstAction);
    store.registerAction("B", mySecondAction);

   // workaround the queue
   (store as any).internalDispatch(myFirstAction);
   (store as any).internalDispatch(mySecondAction);
  }
}
```

Again you'll most likely come into a situation where the state in a parallel running action isn't what you'd expect and you also can't guarantee the order anymore because in above case `mySecondAction` might complete before `firstModification` and the state might not be what it's expected.

---

Bottom line the whole idea of state management is to have a predictable state container. That means you're guaranteed to have a proper:

* Order of states
* Fully reproducible history
* No sideeffects

If you ditch that you most likely don't want to use a state plugin but would be better of with classic services plus events where you know better and handle execution according to your ideas.

**Comment by zewa666** [comment]
There is also a related StackOverflow question (https://stackoverflow.com/questions/54266405/aurelia-change-state-twice-in-one-action/54287249#54287249)

If all you want to do is execute multiple "reducers" in one go you can follow along as the last example depicted

```typescript
function updateRolesAndDisableLoader(state, roles) {
  return Object.assign(
    {},
    updateRoles(state, roles),
    setLoader(state, false)
  );
}
```

That way you apply two actions but still perform one state change.

**Comment by zewa666** [comment]
There is one more way of handling parallel executions in one action

```typescript

const myParallelAction = (state) => {
  return Promise.all([ myservice.asyncActionA(), myservice.asyncActionB() ]).then( (values) => {
    return Object.assign(
     {},
     state,
     { dataFromA: values[0] },
     { dataFromB: values[1] },
  });
}
```

this way both requests are triggered together and only when both complete (notice Promise.all vs Promise.race) the action will be dispatched. Keep in mind though that as long as this action does not resolve or reject the queue is blocked.

**Comment by ConductedClever** [comment]
Great thanks to you (@timfish @zewa666). I have worked many days on your comments and solutions. All of them was right and helpful.
@timfish you are right and it got clear that queuing actions one after another is main dependency of keeping app state predictable. So by now I have had many sessions with our group in company, and we are all accepted that breaking this queue will not be a good idea.
And about @zewa666's answer, these mentions finished my job:
> - Order of states
> - Fully reproducible history
> - No sideeffects

And thank to you for offering other ways to do jobs as our tendency.
And mentioning `Promise.all` was a good solution to reduce waiting problems.

And what I (and others in our team) am doing, is to change this layer architecture:
**VM -> (dispatching) Action(s) -> (calling) Service(s) -> [retrieving from http] -> [updating state by caller Action]**
to this architecture:
**VM -> (calling exactly one) Service (for one job) -> ?(dispatching) Action(s) -> (calling) Service(s) -> [retrieving from http] -> [give data back to first service] -> ?(dispatching) Action(s) (to update state)**
And finally thanks to your quick and keen solutions.
